File: app/src/main/java/com/master/kit/base/BaseFragment.java
package com.master.kit.base;

import android.content.Context;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.master.kit.fragment.DebugFragment;
import com.master.kit.fragment.WidgetFragment;
import com.ooftf.kit.utils.LogUtil;

/**
 * Created by master on 2016/4/12.
 */
public class BaseFragment extends Fragment {
    boolean isVisible;
    boolean isLoaded;
    Runnable doRefresh;
    Toast mToast;
    private boolean isTouchable = false;
    private boolean isAlive = false;
    @Override
    public void onAttach(Context context) {
        isAlive = true;
        LogUtil.e(getClass().getSimpleName(), "onAttach");
        super.onAttach(context);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        LogUtil.e(getClass().getSimpleName(), "onCreate          " + savedInstanceState);
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        LogUtil.e(getClass().getSimpleName(), "onCreateView");
        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        LogUtil.e(getClass().getSimpleName(), "onActivityCreated");
        super.onActivityCreated(savedInstanceState);
    }

    @Override
    public void onStart() {
        LogUtil.e(getClass().getSimpleName(), "onStart");
        super.onStart();
    }

    @Override
    public void onResume() {
        isTouchable = true;
        LogUtil.e(getClass().getSimpleName(), "onResume");
        super.onResume();
        doOnResume();
        loadJudgment();
    }

    @Override
    public void onPause() {
        isTouchable = false;
        LogUtil.e(getClass().getSimpleName(), "onPause");
        super.onPause();
    }

    @Override
    public void onStop() {
        LogUtil.e(getClass().getSimpleName(), "onStop");
        super.onStop();
    }

    @Override
    public void onDestroyView() {
        LogUtil.e(getClass().getSimpleName(), "onDestroyView");
        super.onDestroyView();
    }

    @Override
    public void onDestroy() {
        LogUtil.e(getClass().getSimpleName(), "onDestroy");
        super.onDestroy();
    }

    @Override
    public void onDetach() {
        isAlive = false;
        LogUtil.e(getClass().getSimpleName(), "onDetach");
        super.onDetach();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        LogUtil.e(getClass().getSimpleName(), "onSaveInstanceState     " + outState);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        LogUtil.e(getClass().getSimpleName(), "onHiddenChanged     " + hidden);
        super.onHiddenChanged(hidden);
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        LogUtil.e(getClass().getSimpleName(), "onConfigurationChanged     " + newConfig);
        super.onConfigurationChanged(newConfig);
    }

    private void loadJudgment() {
        if (isTouchable && isVisible && !isLoaded) {
            isLoaded = true;
            onLazyLoad();
        }
    }

    protected void onLazyLoad() {
        //do nothing
    }

    public boolean isTouchable() {
        return isTouchable;
    }

    public boolean isAlive() {
        return isAlive;
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        isVisible = isVisibleToUser;
        LogUtil.e(getClass().getSimpleName(), "setUserVisibleHint");
        super.setUserVisibleHint(isVisibleToUser);
        loadJudgment();
    }

    public void postOnResume(Runnable doRefresh) {
        this.doRefresh = doRefresh;
    }

    private void doOnResume() {
        if (doRefresh != null) {
            doRefresh.run();
            doRefresh = null;
        }
    }

    public void toast(String content) {
        if (mToast != null){
            mToast.cancel();
        }
        mToast = Toast.makeText(getContext(), content, Toast.LENGTH_SHORT);
        mToast.show();
    }

}

File: service/src/main/java/tf/oof/com/service/base/BaseFragment.kt
package tf.oof.com.service.base

import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import tf.oof.com.service.interfaces.ILifecycle
import tf.oof.com.service.utils.LogUtil
import java.util.*

/**
 * Created by master on 2016/4/12.
 */
open class BaseFragment : Fragment(), ILifecycle {
    override fun postOnDestroy(doOnDestroy: () -> Unit) {
        onDestroyList.add(doOnDestroy)
    }

    override fun isAlive(): Boolean {
        return alive
    }

    override fun isShowing(): Boolean {
        return showing
    }

    override fun isTouchable(): Boolean {
        return touchable
    }

    private var isLoaded: Boolean = false
    var mToast: Toast? = null
    private var touchable = false
    private var showing = false
    private var alive = false
    private var onResumeList: MutableList<() -> Unit> = ArrayList()
    private val onDestroyList: MutableList<() -> Unit> by lazy { ArrayList<() -> Unit>() }
    override fun onAttach(context: Context) {
        alive = true
        LogUtil.e(javaClass.simpleName, "onAttach")
        super.onAttach(context)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        LogUtil.e(javaClass.simpleName, "onCreate          " + savedInstanceState)
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        LogUtil.e(javaClass.simpleName, "onCreateView")
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        LogUtil.e(javaClass.simpleName, "onActivityCreated")
        super.onActivityCreated(savedInstanceState)
    }

    override fun onStart() {
        LogUtil.e(javaClass.simpleName, "onStart")
        super.onStart()
    }

    override fun onResume() {
        touchable = true
        LogUtil.e(javaClass.simpleName, "onResume")
        super.onResume()
        doOnResume()
        loadJudgment()
    }

    override fun onPause() {
        touchable = false
        LogUtil.e(javaClass.simpleName, "onPause")
        super.onPause()
    }

    override fun onStop() {
        LogUtil.e(javaClass.simpleName, "onStop")
        super.onStop()
    }

    override fun onDestroyView() {
        LogUtil.e(javaClass.simpleName, "onDestroyView")
        super.onDestroyView()
    }

    override fun onDestroy() {
        LogUtil.e(javaClass.simpleName, "onDestroy")
        super.onDestroy()
    }

    override fun onDetach() {
        alive = false
        doOnDestroy()
        LogUtil.e(javaClass.simpleName, "onDetach")
        super.onDetach()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        LogUtil.e(javaClass.simpleName, "onSaveInstanceState     " + outState)
        super.onSaveInstanceState(outState)
    }

    override fun onHiddenChanged(hidden: Boolean) {
        LogUtil.e(javaClass.simpleName, "onHiddenChanged     " + hidden)
        super.onHiddenChanged(hidden)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        LogUtil.e(javaClass.simpleName, "onConfigurationChanged     " + newConfig)
        super.onConfigurationChanged(newConfig)
    }

    private fun loadJudgment() {
        if (touchable && showing && !isLoaded) {
            isLoaded = true
            onLazyLoad()
        }
    }

    protected fun onLazyLoad() {
        //do nothing
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        showing = isVisibleToUser
        LogUtil.e(javaClass.simpleName, "setUserVisibleHint")
        super.setUserVisibleHint(isVisibleToUser)
        loadJudgment()
    }

    fun postOnResume(doResume: () -> Unit) {
        if(!onResumeList.contains(doResume)){
            onResumeList.add(doResume)
        }
    }

    private fun doOnResume() {
        val iterator = onResumeList.iterator()
        while (iterator.hasNext()) {
            val next = iterator.next()
            next.invoke()
            iterator.remove()
        }
    }

    fun toast(content: String, duration: Int = Toast.LENGTH_SHORT) {
        mToast?.cancel()
        mToast = Toast.makeText(context, content, duration)
        mToast?.show()
    }

    fun getBaseActivity(): BaseActivity {
        return activity as BaseActivity
    }

    private fun doOnDestroy() {
        val iterator = onResumeList.iterator()
        while (iterator.hasNext()) {
            val next = iterator.next()
            next.invoke()
            iterator.remove()
        }
    }
}
